dadb/src/main/kotlin/dadb/forwarding/TcpForwarder.kt

```diff
@@ -0,0 +1,160 @@
+package dadb.forwarding
+
+import dadb.Dadb
+import okio.BufferedSink
+import okio.Source
+import okio.buffer
+import okio.sink
+import okio.source
+import java.io.IOException
+import java.io.InterruptedIOException
+import java.net.ServerSocket
+import java.net.SocketException
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.TimeoutException
+import kotlin.concurrent.thread
+
+internal class TcpForwarder(
+    private val dadb: Dadb,
+    private val hostPort: Int,
+    private val targetPort: Int,
+) : AutoCloseable {
+
+    private var state: State = State.STOPPED
+    private var serverThread: Thread? = null
+    private var server: ServerSocket? = null
+    private var clientExecutor: ExecutorService? = null
+
+    fun start() {
+        check(state == State.STOPPED) { "Forwarder is already started at port $hostPort" }
+
+        moveToState(State.STARTING)
+
+        clientExecutor = Executors.newCachedThreadPool()
+        serverThread = thread {
+            try {
+                handleForwarding()
+            } catch (ignored: SocketException) {
+                // Do nothing
+            } finally {
+                moveToState(State.STOPPED)
+            }
+        }
+
+        waitFor(10, 5000) {
+            state == State.STARTED
+        }
+    }
+
+    private fun handleForwarding() {
+        val serverRef = ServerSocket(hostPort)
+        server = serverRef
+
+        moveToState(State.STARTED)
+
+        while (!Thread.interrupted()) {
+            val client = serverRef.accept()
+
+            clientExecutor?.execute {
```

Comment:
Any reason to use an executor here but create a thread manually for the reader thread?

Reply:
Mainly because of `awaitTermination`

Reply:
Got it